adapters/deeplabv3/model_adapter.py

```diff
@@ -0,0 +1,480 @@
+        self.metrics = {'history': list()}
+        self.model.to(device=self.device)
+
+        ############
+        # Training #
+        ############
+
+        logger.debug("Starting training")
+        for epoch in range(num_epochs):
+            if end_training:
+                break
+            logger.info('Epoch {}/{} Start...'.format(epoch, num_epochs))
+            epoch_time = time.time()
+            # Each epoch has a training and validation phase
+            for phase in ['train', 'val']:
+                if phase == 'train':
+                    self.model.train()  # Set model to training mode
+                else:
+                    self.model.eval()  # Set model to evaluate mode
+
+                running_loss = 0.0
+                running_corrects = 0
+                total = 0
+
+                # Iterate over data
+                with tqdm.tqdm(dataloaders[phase], unit="batch") as tepoch:
+                    for batch in tepoch:
+                        inputs = torch.stack(tuple(batch[0]), 0).to(self.device)
+                        # labels = torch.stack([item['labels'] for item in batch[1]]).to(self.device)
+                        masks = torch.stack([item['masks'] for item in batch[1]]).squeeze(1).to(self.device)
+                        masks = masks.long()
+                        # zero the parameter gradients
+                        optimizer.zero_grad()
+
+                        # forward
+                        with torch.set_grad_enabled(phase == 'train'):
+                            outputs = self.model(inputs)['out']
+                            _, preds = torch.max(outputs, 1)
+
+                            loss = criterion(outputs, masks)
+
+                            # backward + optimize only if in training phase
+                            if phase == 'train':
+                                loss.backward()
+                                optimizer.step()
+
+                        # statistics
+                        total += inputs.size(0)
+                        running_loss += (loss.item() * inputs.size(0))
+                        running_corrects += torch.sum(preds == masks).double().cpu().numpy()
+                        epoch_acc = running_corrects / total
+                        epoch_loss = running_loss / total
+                        tepoch.set_postfix(loss=epoch_loss, accuracy=100. * epoch_acc)
+
+                if phase == 'train':
+                    exp_lr_scheduler.step()
+
+                logger.info(
+                    f'Epoch {epoch}/{num_epochs} - {phase} - Loss: {epoch_loss:.4f}, Acc: {epoch_acc:.4f}, '
+                    f'Duration {(time.time() - epoch_time):.2f}')
+                plots = [dl.PlotSample(figure='loss',
+                                       legend=phase,
+                                       x=epoch,
+                                       y=epoch_loss),
+                         dl.PlotSample(figure='accuracy',
```

Comment:
Doesn't the trainer provide any other figures? Accuracy doesn't seem to be a good metric for the task.